File: Leet-Code-Question/is-graph-bipartite/is-graph-bipartite.java
class Solution {
    int[] parent;
    public boolean isBipartite(int[][] graph) {
        int N = graph.length;
        parent = new int[N];
        for(int i = 0; i < N; i++)
            parent[i] = i;
        
        for(int i = 0; i < N; i++) {
            for(int j = 0; j < graph[i].length; j++) {
                if(find(i) == find(graph[i][j]))
                    return false;
                union(graph[i][0], graph[i][j]);
            }
        }
        
        return true;
    }
    
    private int find(int x) {
        while(x != parent[x])
            x = parent[x];  
        return x;
    }
    
    private void union(int x, int y) {
        int parentX = find(x);
        int parentY = find(y);
        parent[parentY] = parentX;
    }
}

File: Leet-Code-Question/is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        unordered_map<int,bool> m;
        bool currCol=0;
        
        queue<int> q;
        for(int i=0;i<graph.size();i++){
            if(m.find(i)==m.end()){
                        q.push(i);
        m[i]=currCol;
        while(q.size()){
            int node = q.front();
            currCol=m[node];
            q.pop();
            
            for(auto a:graph[node]){
                if(m.find(a)!=m.end() && m[a]==m[node]){
                    return false;
                }else if(m.find(a)==m.end()){
                    m[a]=!currCol;
                    q.push(a);
                }
            }
            // cout<<"\n";
            
        }

            }
        }        
        return true;
    }
};